controllers/configurationpolicy_controller.go

```diff
@@ -130,247 +182,188 @@ type ConfigurationPolicyReconciler struct {
+	if policy.Status.LastEvaluatedGeneration != policy.Generation {
+		r.processedPolicyCache.Delete(policy.GetUID())
+	}
 
-					go r.handlePolicyWorker(policyQueue, &wg)
-				}
+	shouldEvaluate, durationLeft := r.shouldEvaluatePolicy(policy, (uninstalling || crdDeleting))
+	if !shouldEvaluate {
+		// Requeue based on the remaining time for the evaluation interval to be met.
+		return reconcile.Result{RequeueAfter: durationLeft}, nil
+	}
 
-				for i := range policiesList.Items {
-					policy := policiesList.Items[i]
+	before := time.Now().UTC()
 
-					// If the ConfigurationPolicy's spec field was updated, clear the cache of the objects that have
-					// been processed.
-					if policy.Status.LastEvaluatedGeneration != policy.Generation {
-						r.processedPolicyCache.Delete(policy.GetUID())
-					}
+	handleErr := r.handleObjectTemplates(policy)
 
-					if !r.shouldEvaluatePolicy(&policy, cleanupImmediately) {
-						continue
-					}
+	duration := time.Now().UTC().Sub(before)
+	seconds := float64(duration) / float64(time.Second)
 
-					// handle each template in each policy
-					policyQueue <- &policy
-				}
+	policyEvalSecondsCounter.WithLabelValues(policy.Name).Add(seconds)
+	policyEvalCounter.WithLabelValues(policy.Name).Inc()
 
-				close(policyQueue)
-				wg.Wait()
+	if handleErr != nil {
+		// If the policy is invalid, don't bother requeueing since we need to wait for a spec change.
+		if errors.Is(handleErr, ErrPolicyInvalid) {
+			// Remove any watches on the policy in case the policy used to be valid and specified watches.
+			err := r.DynamicWatcher.RemoveWatcher(policy.ObjectIdentifier())
+			if err != nil {
+				log.Error(err, "Failed to remove any watches related to this ConfigurationPolicy. Will ignore.")
 			}
-		}
 
-		// Update the related object metric after policy processing
-		if r.EnableMetrics {
-			updateRelatedObjectMetric()
-		}
-		// Update the evaluation histogram with the elapsed time
-		elapsed := time.Since(start).Seconds()
-		evalLoopHistogram.Observe(elapsed)
-		// making sure that if processing is > freq we don't sleep
-		// if freq > processing we sleep for the remaining duration
-		if float64(freq) > elapsed {
-			remainingSleep := float64(freq) - elapsed
-			sleepTime := time.Duration(remainingSleep) * time.Second
-			log.V(2).Info("Sleeping before reprocessing the configuration policies", "seconds", sleepTime)
-			time.Sleep(sleepTime)
-		}
-
-		select {
-		case <-ctx.Done():
-			exiting = true
-		default:
+			return reconcile.Result{}, nil
 		}
+
+		return reconcile.Result{}, handleErr
 	}
-}
 
-// handlePolicyWorker is meant to be used as a Go routine that wraps handleObjectTemplates.
-func (r *ConfigurationPolicyReconciler) handlePolicyWorker(
-	policyQueue <-chan *policyv1.ConfigurationPolicy, wg *sync.WaitGroup,
-) {
-	defer wg.Done()
+	var requeueAfter time.Duration
 
-	for policy := range policyQueue {
-		before := time.Now().UTC()
+	//nolint: ineffassign,wastedassign
+	err = nil
```

Comment:
It's not clear if/why the linter is wrong here. I might just prefer a new error variable to clear up the confusion - a descriptive name might be all I need to understand it.

Reply:
I agree--this line isn't necessary because everywhere below this it's reassigned before use, so this assignment doesn't need to be here.